data/db-schema/src/main/resources/net/corda/db/schema/config/migration/cpx-creation-v1.0.xml

```diff
@@ -123,38 +111,26 @@
             </column>
         </createTable>
 
-        <addPrimaryKey tableName="cpk_file" columnNames="cpk_name, cpk_version, cpk_signer_summary_hash" constraintName="cpk_file_pk"
+        <addPrimaryKey tableName="cpk_file" columnNames="file_checksum" constraintName="cpk_file_pk"
                        schemaName="${schema.name}"/>
 
-        <addForeignKeyConstraint baseTableName="cpk_file" baseColumnNames="cpk_name, cpk_version, cpk_signer_summary_hash"
-                                 referencedTableName="cpk_metadata" referencedColumnNames="cpk_name, cpk_version, cpk_signer_summary_hash"
+        <addForeignKeyConstraint baseTableName="cpk_file" baseColumnNames="file_checksum"
+                                 referencedTableName="cpk_metadata" referencedColumnNames="file_checksum"
                                  constraintName="FK_cpk_file_cpk_metadata"
                                  baseTableSchemaName="${schema.name}"
                                  referencedTableSchemaName="${schema.name}"/>
 
-        <addUniqueConstraint tableName="cpk_file" columnNames="file_checksum" constraintName="db_cpk_file_uc1"
-                             schemaName="${schema.name}"/>
-
         <createTable tableName="cpk_db_change_log" schemaName="${schema.name}">
-            <column name="cpk_name" type="VARCHAR(255)">
-                <constraints nullable="false"/>
-            </column>
-            <column name="cpk_version" type="VARCHAR(255)">
-                <constraints nullable="false"/>
-            </column>
-            <column name="cpk_signer_summary_hash" type="VARCHAR(255)">
-                <constraints nullable="false"/>
-            </column>
             <column name="cpk_file_checksum" type="VARCHAR(255)">
                 <constraints nullable="false"/>
             </column>
             <column name="file_path" type="VARCHAR(255)">
                 <constraints nullable="false"/>
             </column>
-            <column name="content" type="TEXT">
```

Comment:
I added changeset_id into the primary key of cpk_db_change_log.

Every time the liquibase changelogs are extracted from a CPI, new changelog entities are created with a unique changeset ID.

This may cause some duplication (i.e. if a CPI is force uploaded with the same CPKs as exists already, we'll get duplicate changelogs for those CPKs but with a different changesetId). This simplifies some of the rollback logic.

Reply:
Is this only during force upload?
Isn't this for each CPI upload? E.g. say you have CPI v1 that includes foo.cpk and bar.cpk, then CPI v2 that updates just foo.cpk (but leaves bar.cpk), will it not also create a new (duplicate) entry here?
It may be ok, just feels a bit odd.
A more natural model in this case may be to have a changeset table which keeps a link between changeset_id and cpk_file_checksum where 1 cpk_file_checksum could link to multiple changesets.

Reply:
Updated: I've removed changeset_id completely, we tag changelogs in liquibase with the CPk file checksum. When we need to rollback these changes, we simply get list of CPK file checksums we've applied, load the cpk_db_change_log entries for these checksums, and roll them back, then apply migrations for the new CPI's CPKs.